File: web2pdf.js
#!/usr/bin/env node
const axios = require("axios");
const cheerio = require("cheerio");
const puppeteer = require("puppeteer");
const yargs = require("yargs");
const path = require("path");
var async = require("async");
const fs = require("fs");
const { PDFDocument } = require("pdf-lib");
const urlMap = {};

let scrapeLimit = 100;
let downloadLimit = 10;
let outputDir = "./output";
let outputFileName = "output.pdf";
let baseUrl = "";
let maxDepth = 0;
let runningScraperCounter = 0;
let scrapeUrlQueue = [];
let workingUrls = [];
function sleep(time) {
  return new Promise((resolve) => setTimeout(resolve, time));
}

function ensureDirectoryExists(dirPath) {
  if (!fs.existsSync(dirPath)) {
    fs.mkdirSync(dirPath, { recursive: true });
    console.log(`Directory created: ${dirPath}`);
  } else {
    console.log(`Directory already exists: ${dirPath}`);
  }
}

async function startScrape(url) {
  console.log(`Starting scrape of ${url}`);
  scrapeUrlQueue.push(url);
  while (scrapeUrlQueue.length > 0 || runningScraperCounter > 0) {
    if (runningScraperCounter < scrapeLimit && scrapeUrlQueue.length > 0) {
      url = scrapeUrlQueue.shift();
      scrapeUrl(url);
    } else {
      await sleep(100);
    }
    // console.log(
    //   `scrapeUrlQueue ${scrapeUrlQueue.length} runningScraperCounter ${runningScraperCounter}`
    // );
  }
  console.log("Scraping done found ", workingUrls.length);
}

async function scrapeUrl(url) {
  runningScraperCounter++;
  let htmlContent = "";
  try {
    const { data } = await axios.get(url);
    htmlContent = data;
    workingUrls.push(url);
    const $ = cheerio.load(htmlContent);
    const links = [];

    $("a[href]").each((_, element) => {
      // console.log(`Found link ${$(element).attr("href")}`);

      noAnchor = $(element).attr("href").split("#")[0];
      console.log(noAnchor);
      if (!noAnchor) {
        return;
      }

      let href = noAnchor;

      if (href.startsWith("/")) {
        href = new URL(href, baseUrl).href;
      }
      if (href.startsWith(".")) {
        href = new URL(href, url).href;
      }
      if (!href.startsWith(baseUrl)) {
        // if link starts not with base url, ignore
        return;
      }
      if (href in urlMap) {
        return;
      }
      // console.log(`Adding link ${href}`);
      if (!urlMap[href]) {
        urlMap[href] = true;
        links.push(href);
      }
    });
    scrapeUrlQueue = [...scrapeUrlQueue, ...links];
  } catch (error) {
    console.log(`Error fetching ${url}  `);
  }

  runningScraperCounter--;
}

function urlToFilename(url) {
  return url
    .replace(/^https?:\/\//, "") // Remove protocol
    .replace(/[^a-zA-Z0-9]/g, "_") // Replace non-alphanumeric characters with underscores
    .replace(/__+/g, "_") // Replace multiple underscores with a single underscore
    .replace(/^_|_$/g, ""); // Remove leading and trailing underscores
}

const convertToPdf = async (url, index) => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  await page.goto(url, { waitUntil: "networkidle2" });
  const pdfName = path.join(outputDir, `output_${urlToFilename(url)}.pdf`);
  await page.pdf({ path: pdfName, format: "A4" });
  await browser.close();
  return pdfName;
};

async function mergePdfs(pdfFiles) {
  const mergedPdf = await PDFDocument.create();

  for (const pdfFile of pdfFiles) {
    const pdfBytes = fs.readFileSync(pdfFile);
    const pdf = await PDFDocument.load(pdfBytes);
    const copiedPages = await mergedPdf.copyPages(pdf, pdf.getPageIndices());
    copiedPages.forEach((page) => mergedPdf.addPage(page));
  }

  const mergedPdfBytes = await mergedPdf.save();
  fs.writeFileSync(outputFileName, mergedPdfBytes);
  // delete individual pdfs
  // for (const pdfFile of pdfFiles) {
  //   fs.unlinkSync(pdfFile);
  // }
}

const main = async (startUrl, recursive) => {
  baseUrl = startUrl;
  maxDepth = recursive ? 100 : 0;
  await startScrape(startUrl);
  console.log(`Start downloading page ...`);
  ensureDirectoryExists(outputDir);
  const pdfFiles = await async.mapLimit(
    workingUrls,
    downloadLimit,
    convertToPdf
  );
  console.log(`Combine files to one PDF ...`);
  await mergePdfs(pdfFiles);
  // For PDF merging, you might need to use an external tool or library
  // Code for merging PDFs and cleaning up individual PDFs goes here
};

const argv = yargs(process.argv.slice(2)).options({
  url: {
    type: "string",
    demandOption: true,
    describe: "URL of the webpage to convert",
  },
  recursive: {
    type: "boolean",
    default: false,
    describe: "Recursively follow links",
  },
}).argv;

main(argv.url, argv.recursive);

File: web2pdf.py
import os
import requests
import argparse
from bs4 import BeautifulSoup
from PyPDF2 import PdfFileMerger
from urllib.parse import urlparse, urljoin

import subprocess


def fetch_links(url, visited, base_url):
    print(f"fetch links Url: {url}")
    if url in visited:
        return []
    print(f"Fetching {url}")
    visited.add(url)
    
    # Use requests to get the page content
    page = requests.get(url)
    soup = BeautifulSoup(page.content, 'html.parser')

    links = []
    for a in soup.find_all('a', href=True):

        href = a['href']  # Access href like a dictionary
        if href:
            # Check if the link is relative and join with base URL
            if href.startswith('/'):
                link = urljoin(base_url, href)
                print(base_url+href)
                links.append(link)
            if href.startswith('./'):
                link = urljoin(url, href[2:])  # Remove './' before joining
                links.append(link)
            if href.startswith('http'):
            # Check if the absolute URL belongs to the base domain
                if urlparse(href).netloc == urlparse(base_url).netloc:
                    links.append(link)

    visited.add(url)            
    return links
def convert_to_pdf(url, index, screen_width='1680'):
    pdf_name = f"output_{index}.pdf"
    command = f"wkhtmltopdf --viewport-size {screen_width} {url} {pdf_name}"
    subprocess.run(command, shell=True)
    return pdf_name

def recursive_scrape(url, visited, base_url, max_depth, url_to_pdf, current_depth=0):
    print(f"Url: {url}, Current depth: {current_depth}")
    if current_depth >= max_depth or url in visited:
        return []
    
    # if url not in url_to_pdf:
    #     pdf_name = convert_to_pdf(url, len(url_to_pdf))
    #     url_to_pdf[url] = pdf_name
    # pdf_files = [url_to_pdf[url]]
    pdf_files = []
    links = fetch_links(url, visited, base_url)

    print(links)
    for link in links:
        pdf_files.extend(recursive_scrape(link, visited, base_url, max_depth, url_to_pdf, current_depth + 1))
    return pdf_files


def main(start_url, recursive):
    visited = set()
    max_depth = 100 if recursive else 0
    
    url_to_pdf = {}
    pdf_files = recursive_scrape(start_url, visited, start_url, max_depth, url_to_pdf)

    # Merging PDFs
    merger = PdfFileMerger()
    for pdf in pdf_files:
        merger.append(pdf)
    merger.write("final_output.pdf")
    merger.close()

    # Cleanup individual PDFs
    for pdf in pdf_files:
        os.remove(pdf)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Webpage to PDF converter")
    parser.add_argument("url", type=str, help="URL of the webpage to convert")
    parser.add_argument("-r", "--recursive", action="store_true", help="Recursively follow links")
    args = parser.parse_args()

    main(args.url, args.recursive)
